SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -620,35 +706,227 @@ void ASpatialDebugger::DrawDebug(UCanvas* Canvas, APlayerController* /* Controll
+			{
+				ActorMeshComponents.Add(MeshComponent);
+				// Store previous materials
+				ActorMeshMaterials.Add(MeshMaterial);
+				// Set wireframe material on new actor
+				MeshComponent->SetMaterial(0, WireFrameMaterial);
+			}
+		}
+		HoverActor = NewHoverActor;
+	}
+}
+
+void ASpatialDebugger::RevertHoverMaterials()
+{
+	if (!bShowHighlight)
+	{
+		return;
+	}
+
+	if (HoverActor.IsValid())
+	{
+		// Revert materials on previous actor
+		for (int i = 0; i < ActorMeshComponents.Num(); i++)
+		{
+			TWeakObjectPtr<UMeshComponent> ActorMeshComponent = ActorMeshComponents[i];
+			TWeakObjectPtr<UMaterialInterface> ActorMeshMaterial = ActorMeshMaterials[i];
+			if (ActorMeshComponent.IsValid() && ActorMeshMaterial.IsValid())
+			{
+				ActorMeshComponent->SetMaterial(0, ActorMeshMaterial.Get());
+			}
+		}
+
+		// Clear previous materials
+		ActorMeshMaterials.Empty();
+		ActorMeshComponents.Empty();
+
+		HoverActor = nullptr;
+	}
+}
+
+TWeakObjectPtr<AActor> ASpatialDebugger::GetActorAtPosition(FVector2D& NewMousePosition)
+{
+	if (!LocalPlayerController.IsValid())
+	{
+		return nullptr;
+	}
+	else if (NewMousePosition != MousePosition)
+	{
+		// Mouse has moved so raycast to find actors currently under the mouse cursor
+		MousePosition = NewMousePosition;
+
+		FVector WorldLocation;
+		FVector WorldRotation;
+		LocalPlayerController->DeprojectScreenPositionToWorld(NewMousePosition.X, NewMousePosition.Y, WorldLocation,
+															  WorldRotation); // Mouse cursor position
+		FVector StartTrace = WorldLocation;
+		FVector EndTrace = StartTrace + WorldRotation * MaxRange;
+
+		HitActors.Empty();
+
+		TArray<FHitResult> HitResults;
+		bool bHit = GetWorld()->LineTraceMultiByObjectType(HitResults, StartTrace, EndTrace, CollisionObjectParams);
+		if (bHit)
+		{
+			// When the raycast hits an actor then it is highlighted, whilst the actor remains under the crosshair. If there are multiple
+			// hit results, the user can select the next by using the mouse scroll wheel
+			for (const FHitResult& HitResult : HitResults)
+			{
+				const TWeakObjectPtr<AActor> HitActor = HitResult.GetActor();
+
+				if (!HitActor.IsValid() || HitActors.Contains(HitActor))
 				{
+					// The hit results may include the same actor multiple times so just ignore duplicates
 					continue;
 				}
 
-				DrawTag(Canvas, ScreenLocation, EntityId, Actor->GetName(), true /*bCentre*/);
+				if (const Worker_EntityId_Key* HitEntityId = EntityActorMapping.FindKey(HitResult.GetActor()))
+				{
+					FVector PlayerLocation = GetLocalPawnLocation();
+
+					FVector2D ScreenLocation = ProjectActorToScreen(HitActor, PlayerLocation);
+					if (!ScreenLocation.IsZero())
+					{
+						HitActors.Add(HitActor);
+					}
+				}
 			}
 		}
+		ValidateHoverIndex();
+	}
+
+	// Return actor selected from list dependent on the hover index, which is selected independently with the mouse wheel
+	if (HitActors.Num() > 0)
```

Comment:
```suggestion
	if (HitActors.Num() > HoverIndex)
```

Reply:
Encapsulated this behaviour into a GetHitActor that validates the hover index and checks the number of actors before indexing the array.